README.md
# SlowMotionServo

A Slow Motion Servo Library for Arduino

SlowMotionServo provides a way to drive one or many servos slowly. Instead of
setting the position, each servo has 2 trajectories, ie 2 functions angle(t)
where t is the time. The first function is used to compute the angle of the
servo when travelling from minimum to maximum angle and the second one is used
when travelling from the maximum to the minimum angle.

The function is normalized between 0.0 and 1.0, both for the angle and the time.
So the angle varies from the minimum angle set for the servo to the maximum
angle set for the servo so that the minimum angle corresponds to 0.0 and the
maximum angle corresponds to 1.0. time 0.0 is the start of the movement and time
1.0 is the end of the movement. By default the speed is set so that travelling
from the minimum angle to the maximum angle takes 10s. The same exists when
travelling from the maximum to the minimum angle and both travels have their
own speed setting.

3 trajectories are available and you can add more trajectories by deriving from
the class SlowMotionServo:
* linear trajectory. angle = t
* sinusoidal trajectory. angle = 1.0 - cos(t * PI))/2.0
* sinusoidal trajectory with a bounce. When t <= 0.79, angle = 1.0 - cos(t * PI))/1.8. Otherwise, timeOff = 10.0 * (t - 0.55) and angle = 0.834 + 1.0 / (timeOff * timeOff)

## Using the library

As usual, you have to include it at the beginning of your sketch but you also
have to include the Servo library:

```
#include <Servo.h>
#include <SlowMotionServo.h>
```

Then you have to instantiate as many object as servos you want to drive. For
that you choose what kind of trajectory you want:

SMSLinear is the class with a linear trajectory. SMSSmooth is the class with a
sinusoidal trajectory and SMSSmoothBounce is the class with a sinusoidal
trajectory with a bounce.

```
SMSLinear myServo; /* Servo with linear trajectory */
```

The following functions are available:

### setMin(min)

Set the minimum angle of the servo. The angle is expressed in its equivalency
in microseconds. The value can range from 544 to 2400. A value lower than 544
will be reset to 544 and a value greater than 2400 will be reset to 2400.
If the value is greater than the maximum angle it is reset to the maximum
angle.

### setMax(max)

Set the maximum angle of the servo. The angle is expressed in its equivalency
in microseconds. The value can range from 544 to 2400. A value lower than 544
will be reset to 544 and a value greater than 2400 will be reset to 2400.
If the value is lower than the minimum angle it is reset to the minimum
angle.

### setMinMax(min, max)

Set the minimum and maximum angles of the servo. The angle is expressed in
its equivalency in microseconds. The value can range from 544 to 2400.
A value lower than 544 will be reset to 544 and a value greater than 2400
will be reset to 2400.
If the minimum angle is greater than the maximum angle, both angles are set
to the average value. For instance if you set the minimum angle to 2000 and
the maximum angle to 1000, both angles will be set to 1500.

### setPin(pin)

Set the pin to which the servo is attached.

### setMinToMaxSpeed(speed)

Set the speed of the servo when travelling from the minimum to the maximum
angle. speed is a floating point number. A speed of 1.0 corresponds to a
10s travelling.

### setMaxToMinSpeed(speed)

Set the speed of the servo when travelling from the maximum to the minimum
angle. speed is a floating point number. A speed of 1.0 corresponds to a
10s travelling.

### setSpeed(speed)

Set the speed of the servo when travelling from the minimum to the maximum
angle and from the maximum to the minimum angle.
speed is a floating point number. A speed of 1.0 corresponds to a
10s travelling.

### setInitialPosition(position)

Set the initial position of the servo. The position is a floating point number
ranging from 0.0 to 1.0. If the value is greater than 1.0, ti is reset to 1.0
and if lower than 0.0, it is reset to 0.0

### setReverted(reverted)

Reverse the movement. By default reverted is false. If set to true, the trajectories are mirrored across an axis at time = 0.5.

### goTo(position)

Go to the specified position by following the trajectory.
The position is a floating point number
ranging from 0.0 to 1.0. If the value is greater than 1.0, ti is reset to 1.0
and if lower than 0.0, it is reset to 0.0

### goToMin()

Equivalent to goTo(0.0)

### goToMax()

Equivalent to goTo(1.0)

### setDetachAtMin(detach)

detach is a boolean. If true, the servo is detached when the minimum position
is reached. The servo is no longer driven. This is useful when the servo has
to push against an elastic restoring force. If false the servo continues to be
driven.

### setDetachAtMax(detach)

detach is a boolean. If true, the servo is detached when the maximum position
is reached. The servo is no longer driven. This is useful when the servo has
to push against an elastic restoring force. If false the servo continues to be
driven.

### setDetach(detach)

detach is a boolean. If true, the servo is detached when the minimum or the
maximum positions
are reached. The servo is no longer driven. This is useful when the servo has
to push against an elastic restoring force. If false the servo continues to be
driven.

### isStopped()

Returns true if the servo is stopped.

### pin()

Returns a ```byte``` which is the number of the pin. 255 is returned if the object
has not been connected to any pin.

### detachAtMin()

Returns true if the servo is detached when reaching the minimum position.

### detachAtMax()

Returns true if the servo is detached when reaching the maximum position.

### minimumPulse()

Returns an ```uint16_t``` which is the pulse width in microseconds corresponding to the minimum servo position.

### maximumPulse()

Returns an ```uint16_t``` which is the pulse width in microseconds corresponding to the maximum servo position.

### minToMaxSpeed()

Returns a float which is the speed of the servo when traveling from minimum to maximum position.

### maxToMinSpeed()

Returns a float which is the speed of the servo when traveling from maximum to minimum position.

### isReverted()

Returns true if the movement is reverted.

### SlowMotionServo::setDelayUntilStop(delay)

This class function set the delay between the time the servos reach their
minimum or maximum
position and the the time they are detached. Because the mechanic is always
late
compared to the program, detaching immediately the servos would prevent them
to reach their mechanical position. This is set once for all the servos and
used only for servos and positions for which setDetach(true) is specified.

### SlowMotionServo::update()

Update the positions of all the servos. This class function has to be called
in loop() as frequently as possible.

## Live adjustement of minimum and maximum position

With version 1.1, 5 functions have been added to implement live adjustement of minimum and maximum positions. When an adjustement of a servo is in progress, the goTo, goToMin and goToMax functions fail silently.

Entering in live adjustement process is done implicitely by calling one of the for adjust functions. Exiting the live adjustement process is done by calling de ```endAdjust()``` function.

### setupMin(minPulse)

```setupMin()``` sets the minimum pulse value. ```minPulse``` is an unsigned 16 bits integer. The servo enters in the live adjustement process: its state is saved, the minimum pulse is set to ```minPulse``` and it goes to the new minimum position so that the user can see it. If ```minPulse``` is greater than the maximum pulse, it is set back to the maximum pulse. If ```minPulse``` is lower than the minimum possible value of the Servo library (544), it is set back to the minimum possible value.

### setupMax(maxPulse)

```setupMax()``` sets the maximum pulse value. ```maxPulse``` is an unsigned 16 bits integer. The servo enters in the live adjustement process: its state is saved, the maximum pulse is set to ```maxPulse``` and it goes to the new maximum position so that the user can see it. If ```maxPulse``` is lower than the minimum pulse, it is set back to the minimum pulse. If ```maxPulse``` is greater than the maximum possible value of the Servo library (2400), it is set back to the maximum possible value.

### adjustMin(increment)

```adjustMin()``` adds ```increment``` to the minimum position. ```increment``` is a signed 16 bits integer. The servo enters in the live adjustement process: its state is saved, ```increment``` is added to the minimum pulse and it goes to the new minimum position so that the user can see it. If after adding ```increment```, the minimum pulse is greater than the maximum pulse, it is set back to the maximum pulse. If after adding ```increment```, the minimum pulse is lower than the minimum possible value of the Servo library (544), it is set back to the minimum possible value.

### adjustMax(increment)

```adjustMax()``` adds ```increment``` to the maximum position. ```increment``` is a signed 16 bits integer. The servo enters in the live adjustement process: its state is saved, ```increment``` is added to the maximum pulse and it goes to the new maximum position so that the user can see it. If after adding ```increment```, the maximum pulse is lower than the minimum pulse, it is set back to the minimum pulse.  If after adding ```increment```, the maximum pulse is greater than the maximum possible value of the Servo library (2400), it is set back to the maximum possible value.

### endSetup()

After calling the 4 previous function, ```endSetup()``` shall be called to exit the adjustement process. The state of the servo is restored. However it does not returns to the position it was before the adjustement process.

## Example

Let's drive a single servo with a smooth movement forward and backward.

First, include the libraries:

```
#include <Servo.h>
#include <SlowMotionServo.h>
```

Second, instantiate an object and a float to hold the target position:

```
SMSSmooth myServo; /* Servo with linear trajectory */
float target = 0.0;
```

Third, initialize it in setup(). Be careful actual minimum and maximum
positions of a servo may be greater and/or lower than positions allowed by
the Servo and SlowMotionServo libraries:

```
void setup()
{
  myServo.setInitialPosition(target);
  myServo.setMinMax(700, 2000);
  myServo.setPin(3); /* the servo is connected to pin 3 */
}
```

Fourth, if the servo is stopped compute the new target position by doing
the subtraction 1.0 - target so that if target is 0.0, the new target is
1.0 and if target is 1.0 the new target is 0.0. And of course call update.

```
void loop()
{
  if (myServo.isStopped()) {
    target = 1.0 - target;
    myServo.goTo(target);
  }

  SlowMotionServo::update();
}
```

The full sketch:

```
#include <Servo.h>
#include <SlowMotionServo.h>

SMSSmooth myServo; /* Servo with linear trajectory */
float target = 0.0;

void setup() {
  myServo.setInitialPosition(target);
  myServo.setMinMax(700, 2000);
  myServo.setPin(3); /* the servo is connected to pin 3 */
}

void loop() {
  if (myServo.isStopped()) {
    target = 1.0 - target;
    myServo.goTo(target);
  }

  SlowMotionServo::update();
}
```

## How to define your own trajectory.

To do that, you have to inherit from the SlowMotionServo class and redefine
the slopeUp and slopeDown member functions. Let's take the
SMSSmooth class as example:

```
class SMSSmooth : public SlowMotionServo
{
public:
  virtual float slopeUp(float time);
  virtual float slopeDown(float time);
  float slope(float time) { return (1.0 - cos(time * PI))/2.0; }
};
```

Because trajectories are the same from min to max and max to min, we define
a new member function, slope, that defines the trajectory. This function is
called by slopeUp and slopeDown:

```
float SMSSmooth::slopeUp(float time)
{
  return slope(time);
}

float SMSSmooth::slopeDown(float time)
{
  return slope(time);
}
```


changelog.txt
+---------------------------------------------------------------------------+
| SlowMotionServo changelog                                                 |
+---------------------------------------------------------------------------+
1.1.2   Fix a mistake in the documentation.
1.1.1   Live settings example
1.1     Added methods to support live min max settings
1.0.6   Release to fix the problem of the duplicated 1.0.5. Nothing new
1.0.5   Minor fix in one of the accessor name
1.0.4   Added accessors to get almost all attributes of a SMS object
1.0.3   Added a method to do a mirror at time = .5
1.0.2   Default min and max have been changed to 1000 and 2000.
        More accurate SMSSmoothBounce trajectory.
				Fixed a problem with the initial position.
        Added PushButton example.
1.0.1   Fix a typo in the library.properties
1.0     Initial release.


src/SlowMotionServo.h
/*
 * Slow Motion Servo Library for Arduino
 *
 * Copyright Jean-Luc Béchennec 2015
 *
 * This software is distributed under the GNU Public Licence v2 (GPLv2)
 *
 * Please read the LICENCE file
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#ifndef __SlowMotionServo_h__
#define __SlowMotionServo_h__

#include "Arduino.h"
#include <Servo.h>

class SlowMotionServo : public Servo
{
private:
  byte mPin:6;                // connection pin 0..63
  bool mDetachAtMin:1;
  bool mDetachAtMax:1;
  byte mState:3;              // state of the servo
  byte mSavedState:3;         // saved state of the servo for setup
  bool mReverted:1;
  unsigned int mMinPulse;     // minimum position of the servo in microseconds
  unsigned int mMaxPulse;     // maximum position of the servo in microseconds
  unsigned long mStartTime;   // time when the movement begin
  float mInitialRelativeTime; // starting position-time of the servo
  float mTargetRelativeTime;  // target position-time of the servo
  float mCurrentRelativeTime; // current position-time of the servo
  float mTimeFactorUp;        // time factor of a millis for movement from min to max
  float mTimeFactorDown;      // time factor of a millis for movement from max to min
  SlowMotionServo *mNext;     // next servo in the list

  static SlowMotionServo *sServoList; // head of the servo list
  static unsigned int sDelayUntilStop;

  void updatePosition();      // update the position of the servo
  void updatePulseAccordingToMinMax();
  unsigned int normalizePos(const unsigned int inPos);
  bool isSettable();

public:
  SlowMotionServo();
  SlowMotionServo(byte pin);
  void setPin(const byte pin) { mPin = pin; }
  void setMinMax(unsigned int minPulse, unsigned int maxPulse);
  void setMin(unsigned int minPulse);
  void setMax(unsigned int maxPulse);
  unsigned int minPosition() { return mMinPulse; }
  unsigned int maxPosition() { return mMaxPulse; }
  void setReverted(const bool inReverted);
  void setMinToMaxSpeed(const float speed);
  void setMaxToMinSpeed(const float speed);
  void setSpeed(const float speed);
  void setInitialPosition(float position);
  void setDetachAtMin(bool detach) { mDetachAtMin = detach; }
  void setDetachAtMax(bool detach) { mDetachAtMax = detach; }
  void setDetach(bool detach) { mDetachAtMin = mDetachAtMax = detach; }
  void goTo(float position);
  void goToMin() { goTo(0.0); }
  void goToMax() { goTo(1.0); }
  bool isStopped();

  /*
   * Live configuration functions.
   */
  void setupMin(uint16_t minPulse);
  void setupMax(uint16_t maxPulse);
  void adjustMin(int16_t increment) { setupMin(mMinPulse + increment); }
  void adjustMax(int16_t increment) { setupMax(mMaxPulse + increment); }
  void endSetup();

  byte pin()                  { return mPin; }
  bool detachAtMin()          { return mDetachAtMin; }
  bool detachAtMax()          { return mDetachAtMax; }
  unsigned int minimumPulse() { return mMinPulse; }
  unsigned int maximumPulse() { return mMaxPulse; }
  float minToMaxSpeed();
  float maxToMinSpeed();
  bool isReverted()           { return mReverted; }

  virtual float slopeUp(float time) = 0;
  virtual float slopeDown(float time) = 0;

  static void update();
  static void setDelayUntilStop(unsigned int delayUntilStop);
};

class SMSLinear : public SlowMotionServo
{
public:
  virtual float slopeUp(float time);
  virtual float slopeDown(float time);
  float slope(float time) { return time; }
};

class SMSSmooth : public SlowMotionServo
{
public:
  virtual float slopeUp(float time);
  virtual float slopeDown(float time);
  float slope(float time) { return (1.0 - cos(time * PI))/2.0; }
};

class SMSSmoothBounce : public SlowMotionServo
{
public:
  virtual float slopeUp(float time);
  virtual float slopeDown(float time);
};

#endif


src/SlowMotionServo.cpp
/*
 * Slow Motion Servo Library for Arduino
 *
 * Copyright Jean-Luc Béchennec 2015
 *
 * This software is distributed under the GNU Public Licence v2 (GPLv2)
 *
 * Please read the LICENCE file
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include <SlowMotionServo.h>

static const byte NOPIN = 63;

enum {
	SERVO_INIT         = 0,
	SERVO_STOPPED      = 1,
	SERVO_SETUP        = 2,
	SERVO_UP           = 3,
	SERVO_DOWN         = 4,
	SERVO_DELAYED_UP   = 5,
	SERVO_DELAYED_DOWN = 6,
	SERVO_NO_STATE     = 7
};

/*
 * Static member to store a list of servos. This allow to update the
 * positions of all declared servos easily.
 */
SlowMotionServo *SlowMotionServo::sServoList = NULL;

/*
 * Delay between the last pulse chage in a movement and the time the
 * servos go to the SERVO_STOPPED state.
 * 10 ms by default. Can be changed by SlowMotionServo::setDelayUntilStop
 */
unsigned int SlowMotionServo::sDelayUntilStop = 10;

/*
 * Constructor :
 * Set the connection pin at NOPIN.
 * The state is SERVO_STOPPED.
 * The servo is not detached by default when stopped.
 * Min and max pulse are set according to min and max value of the servo library.
 * Servo is at the min.
 * Speed are set to 1 (mTimeFactorUp and mTimeFactorDown).
 * The servo is added to the list.
 */
SlowMotionServo::SlowMotionServo() :
  mPin(NOPIN),
	mDetachAtMin(false),
  mDetachAtMax(false),
	mState(SERVO_INIT),
	mSavedState(SERVO_NO_STATE),
  mReverted(false),
  mMinPulse(1000),
  mMaxPulse(2000),
  mInitialRelativeTime(0.0),
  mTargetRelativeTime(0.0),
  mCurrentRelativeTime(0.0),
  mTimeFactorUp(0.0001),
  mTimeFactorDown(0.0001)
{
  mNext = sServoList;
  sServoList = this;
}

/*
 * Constructor with the pin
 */
SlowMotionServo::SlowMotionServo(byte pin) :
  SlowMotionServo()
{
  mPin = pin;
}

/*
 * Function to constrain a pulse between the values allowed
 * by the servo library
 */
static unsigned int constrainPulse(unsigned int pulse)
{
  if (pulse < 544)  pulse = 544;
  else if (pulse > 2400) pulse = 2400;
  return pulse;
}

/*
 * Function to constrain a position between 0.0 and 1.0
 */
static float constrainPosition(float position)
{
  if (position > 1.0) position = 1.0;
  else if (position < 0.0) position = 0.0;
  return position;
}

/*
 * Method to update the current pulse of the servo
 * according to the min and max
 */
void SlowMotionServo::updatePulseAccordingToMinMax()
{
  unsigned int currentPosition = readMicroseconds();
  if (currentPosition < mMinPulse) currentPosition = mMinPulse;
  else if (currentPosition > mMaxPulse) currentPosition = mMaxPulse;
  writeMicroseconds(currentPosition);
}

/*
 * Revert the pulse (mirror at time = .5)
 */
unsigned int SlowMotionServo::normalizePos(const unsigned int inPos)
{
  return mReverted ? map(inPos, mMinPulse, mMaxPulse, mMaxPulse, mMinPulse) : inPos;
}

/*
 * Change the min and max. Constrain both. Check min <= max.
 * If not min and max are sit to the average of both.
 */
void SlowMotionServo::setMinMax(unsigned int minPulse, unsigned int maxPulse)
{
	if (isSettable()) {
		minPulse = constrainPulse(minPulse);
		maxPulse = constrainPulse(maxPulse);
		if (minPulse <= maxPulse) {
		  mMinPulse = minPulse;
		  mMaxPulse = maxPulse;
		}
		else {
		  mMinPulse = mMaxPulse = (minPulse + maxPulse) / 2;
		}
		updatePulseAccordingToMinMax();
	}
}

/*
 * Change the min. Constrain it. Check min <= max.
 * If not min is set to max.
 */
void SlowMotionServo::setMin(unsigned int minPulse)
{
	if (isSettable()) {
	  minPulse = constrainPulse(minPulse);
	  if (minPulse > mMaxPulse) minPulse = mMaxPulse;
	  mMinPulse = minPulse;
	  updatePulseAccordingToMinMax();
	}
}

/*
 * Change the max. Constrain it. Check min <= max.
 * If not max is set to min.
 */
void SlowMotionServo::setMax(unsigned int maxPulse)
{
	if (isSettable()) {
	  maxPulse = constrainPulse(maxPulse);
	  if (maxPulse < mMinPulse) maxPulse = mMinPulse;
	  mMaxPulse = maxPulse;
	  updatePulseAccordingToMinMax();
	}
}

/*
 * if inReverted is true, the movement is reverted. Otherwise it is not
 */
void SlowMotionServo::setReverted(const bool inReverted)
{
	if (isSettable()) {
		mReverted = inReverted;
	}
}

/*
 * Set the speed when travelling from minimum position to maximum positions
 */
void SlowMotionServo::setMinToMaxSpeed(const float speed)
{
	if (isSettable()) {
		mTimeFactorUp = speed / 10000.0;
	}
}

/*
 * Set the speed when travelling from maximum position to minimum positions
 */
void SlowMotionServo::setMaxToMinSpeed(const float speed)
{
	if (isSettable()) {
		mTimeFactorDown = speed / 10000.0;
	}
}

/*
 * Set the speed when travelling from minimum position to maximum positions and
 * from the maximum to minimum position.
 */
void SlowMotionServo::setSpeed(const float speed)
{
	if (isSettable()) {
		mTimeFactorUp = mTimeFactorDown = speed / 10000.0;
	}
}

/*
 * Set the initial position of the servo. The initial position
 * is supposed to be retrieved from the EEPROM or from
 * an other mean
 */
void SlowMotionServo::setInitialPosition(float position)
{
	if (isSettable()) {
	  position = constrainPosition(position);
	  mTargetRelativeTime = mInitialRelativeTime = mCurrentRelativeTime = position;
	}
}

/*
 * Go to a new position. Compute the direction and change the state
 * of the servo. Attach it if the pin is set.
 */
void SlowMotionServo::goTo(float position)
{
  position = constrainPosition(position);
  mTargetRelativeTime = position;
  mInitialRelativeTime = mCurrentRelativeTime;
  if (mTargetRelativeTime > mInitialRelativeTime) {
    mState = SERVO_UP;
    if (mPin != NOPIN) attach(mPin);
  }
  else if (mTargetRelativeTime < mInitialRelativeTime) {
    mState = SERVO_DOWN;
    if (mPin != NOPIN) attach(mPin);
  }
  else mState = SERVO_STOPPED;
  mStartTime = millis();
}

/*
 * Update the position of the servo.
 */
void SlowMotionServo::updatePosition()
{
  float position;
  unsigned long date = millis();

  switch (mState) {
    case SERVO_INIT:
    	position = slopeUp(mCurrentRelativeTime);
      writeMicroseconds(normalizePos(position * (mMaxPulse - mMinPulse) + mMinPulse));
      mState = SERVO_STOPPED;
      break;
    case SERVO_UP:
      mCurrentRelativeTime = (float)(date - mStartTime) * mTimeFactorUp +
                             mInitialRelativeTime;
      if (mCurrentRelativeTime > mTargetRelativeTime) {
        mCurrentRelativeTime = mTargetRelativeTime;
        mState = SERVO_DELAYED_UP;
      }
      position = slopeUp(mCurrentRelativeTime);
      writeMicroseconds(normalizePos(position * (mMaxPulse - mMinPulse) + mMinPulse));
      break;
    case SERVO_DOWN:
      mCurrentRelativeTime = mInitialRelativeTime -
                             (float)(date - mStartTime) * mTimeFactorDown;
      if (mCurrentRelativeTime < mTargetRelativeTime) {
        mCurrentRelativeTime = mTargetRelativeTime;
        mState = SERVO_DELAYED_DOWN;
      }
      position = slopeDown(mCurrentRelativeTime);
      writeMicroseconds(normalizePos(position * (mMaxPulse - mMinPulse) + mMinPulse));
      break;
    case SERVO_DELAYED_UP:
      if ((date - mStartTime) > sDelayUntilStop) {
        mState = SERVO_STOPPED;
        if (mDetachAtMax) detach();
        mStartTime = date;
      }
      break;
    case SERVO_DELAYED_DOWN:
      if ((date - mStartTime) > sDelayUntilStop) {
        mState = SERVO_STOPPED;
        if (mDetachAtMin) detach();
        mStartTime = date;
      }
      break;
  }
}

/*
 * Returns true if the servo is isStopped
 */
bool SlowMotionServo::isStopped()
{
	return (mState == SERVO_STOPPED);
}

/*
 * Returns true if attributes can be set. That means the servo is in the
 * SERVO_INIT state or in the SERVO_STOPPED state or in the SERVO_SETUP STATED
 */
bool SlowMotionServo::isSettable()
{
	return (mState <= SERVO_SETUP);
}

/*
 * setup the minimum position in Live
 */
void SlowMotionServo::setupMin(uint16_t minPulse)
{
	if (isSettable()) {
		if (mState != SERVO_SETUP) {
			mSavedState = mState;
			if (! attached()) attach(mPin);
		}
		mState = SERVO_SETUP;
		minPulse = constrainPulse(minPulse);
		if (minPulse > mMaxPulse) minPulse = mMaxPulse;
		mMinPulse = minPulse;
		writeMicroseconds(minPulse);
	}
}

/*
 * setup the maximum position in Live
 */
void SlowMotionServo::setupMax(uint16_t maxPulse)
{
	if (isSettable()) {
		if (mState != SERVO_SETUP) {
			mSavedState = mState;
			if (! attached()) attach(mPin);
		}
		mState = SERVO_SETUP;
		maxPulse = constrainPulse(maxPulse);
		if (maxPulse < mMinPulse) maxPulse = mMinPulse;
		mMaxPulse = maxPulse;
		writeMicroseconds(maxPulse);
	}
}

void SlowMotionServo::endSetup()
{
	if (mState == SERVO_SETUP) {
		mState = mSavedState;
		if (isStopped()) {
			if ((mCurrentRelativeTime == 0.0 && mDetachAtMin) ||
			    (mCurrentRelativeTime == 1.0 && mDetachAtMax))
			{
				detach();
			}
		}
	}
}


/*
 * Class method to update all the servos
 */
void SlowMotionServo::update()
{
  SlowMotionServo *servo = sServoList;
  while (servo != NULL) {
    servo->updatePosition();
    servo = servo->mNext;
  }
}

float SlowMotionServo::minToMaxSpeed()
{
	return 10000.0 * mTimeFactorUp;
}

float SlowMotionServo::maxToMinSpeed()
{
	return 10000.0 * mTimeFactorDown;
}

/*
 * Set the delay between the end of a movement and the time the
 * servo is stopped and detached if needed.
 */
void SlowMotionServo::setDelayUntilStop(unsigned int delayUntilStop)
{
  sDelayUntilStop = delayUntilStop;
}

float SMSLinear::slopeUp(float time)
{
  return slope(time);
}

float SMSLinear::slopeDown(float time)
{
  return slope(time);
}

float SMSSmooth::slopeUp(float time)
{
  return slope(time);
}

float SMSSmooth::slopeDown(float time)
{
  return slope(time);
}

float SMSSmoothBounce::slopeUp(float time)
{
  if (time <= 0.795) {
    return (1.0 - cos(time * PI))/1.8;
  }
  else {
    float timeOff = 10.0 * (time - 0.55);
    return (0.83 + 1.0 / (timeOff * timeOff));
  }
}

float SMSSmoothBounce::slopeDown(float time)
{
  return (1.0 - cos(time * PI))/2.265;
}
